File: frontend/src/api/process.ts
import * as Processipc from "../../bindings/voxesis/src/Communication/InterProcess/processipc"
import {envIsWails} from "./common";
import {ProcessType} from "../../bindings/voxesis/src/Common/Manager";
import { Events } from "@wailsio/runtime";

export async function GetProcessStatus(uuid: number) {
    if (envIsWails) {
        return Processipc.GetProcessStatus(uuid)
    } else {
        const res = await fetch("/api/process/GetProcessStatus", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                uuid: uuid,
            })
        })

        return res.json()
    }
}

export async function NewProcess(processType: ProcessType, abs: boolean, relPath: string, ...args: string[]) {
    if (envIsWails) {
        return Processipc.NewProcess(processType, abs, relPath, ...args)
    } else {
        const res = await fetch("/api/process/NewProcess", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                processType: processType,
                abs: abs,
                relPath: relPath,
                args: args
            })
        })

        return res.json()
    }
}

export async function SendCommand(uuid: number, command: string) {
    if (envIsWails) {
        return Processipc.SendCommand(uuid, command)
    } else {
        const res = await fetch("/api/process/SendCommand", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({uuid, command})
        })

        return res.json()
    }
}

export async function Start(uuid: number) {
    if (envIsWails) {
        return Processipc.Start(uuid)
    } else {
        const res = await fetch("/api/process/Start", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({uuid})
        })

        return res.json()
    }
}

export async function Stop(uuid: number) {
    if (envIsWails) {
        return Processipc.Stop(uuid)
    } else {
        const res = await fetch("/api/process/Stop", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({uuid})
        })

        return res.json()
    }
}

export async function GetProcessOutput(uuid: number, callback: (data: string) => void) {
    if (envIsWails) {
        Events.On("process-" + uuid + "-output", (data) => {
            callback(data.data);
        });
    } else {
        const ws = new WebSocket("ws://localhost:8080/api/process/GetProcessOutput")
        ws.onmessage = (event) => {
            const data = JSON.parse(event.data)
            if (data.uuid == uuid) {
                callback(data.data)
            }
        }
    }
}

export default {
    GetProcessStatus,
    NewProcess,
    SendCommand,
    Start,
    Stop
}

File: src/Communication/InterHttp/Controller/process.go
package v_web_controller

import (
	"net/http"
	vlogger "voxesis/src/Common/Logger"
	vmanager "voxesis/src/Common/Manager"
	communication "voxesis/src/Communication"

	"github.com/gin-gonic/gin"
	"github.com/gorilla/websocket"
)

type Process struct {
	ws *websocket.Conn
}

func (p *Process) NewProcess(context *gin.Context) {
	var data map[string]interface{}

	if err := context.ShouldBindJSON(&data); err != nil {
		vlogger.AppLogger.Error(err.Error())
		context.JSON(400, nil)
		return
	}

	processType, ok := data["processType"].(float64)
	if !ok {
		context.JSON(400, []interface{}{nil, "invalid processType type"})
		return
	}

	abs, ok := data["abs"].(bool)
	if !ok {
		context.JSON(400, []interface{}{nil, "invalid abs type"})
		return
	}

	relPath, ok := data["relPath"].(string)
	if !ok {
		context.JSON(400, []interface{}{nil, "invalid relPath type"})
		return
	}

	var args []string
	if argsData, exists := data["args"]; exists && argsData != nil {
		argsInterface, ok := argsData.([]interface{})
		if !ok {
			context.JSON(400, []interface{}{nil, "invalid args type"})
			return
		}

		args = make([]string, len(argsInterface))
		for i, arg := range argsInterface {
			argStr, ok := arg.(string)
			if !ok {
				context.JSON(400, []interface{}{nil, "invalid argument type in args"})
				return
			}
			args[i] = argStr
		}
	}

	uuid := communication.ProcessIpc.NewProcess(vmanager.ProcessType(processType), abs, relPath, args...)

	context.JSON(200, uuid)
}

func (p *Process) Start(context *gin.Context) {
	var data map[string]interface{}

	if err := context.ShouldBindJSON(&data); err != nil {
		context.JSON(400, err.Error())
		return
	}

	uuid, ok := data["uuid"].(float64)
	if !ok {
		context.JSON(400, "invalid uuid type")
		return
	}

	err := communication.ProcessIpc.Start(int(uuid))
	if err != nil {
		context.JSON(400, *err)
		return
	}
	context.JSON(200, nil)
}

func (p *Process) Stop(context *gin.Context) {
	var data map[string]interface{}

	if err := context.ShouldBindJSON(&data); err != nil {
		context.JSON(400, err.Error())
		return
	}

	uuid, ok := data["uuid"].(float64)
	if !ok {
		context.JSON(400, "invalid uuid type")
		return
	}

	err := communication.ProcessIpc.Stop(int(uuid))
	if err != nil {
		context.JSON(400, *err)
		return
	}
	context.JSON(200, nil)
}

func (p *Process) SendCommand(context *gin.Context) {
	var data map[string]interface{}

	if err := context.ShouldBindJSON(&data); err != nil {
		context.JSON(400, err.Error())
		return
	}

	uuid, ok := data["uuid"].(float64)
	if !ok {
		context.JSON(400, "invalid uuid type")
		return
	}

	err := communication.ProcessIpc.SendCommand(int(uuid), data["command"].(string))
	if err != nil {
		context.JSON(400, *err)
		return
	}
	context.JSON(200, nil)
}

func (p *Process) GetProcessStatus(context *gin.Context) {
	var data map[string]interface{}

	if err := context.ShouldBindJSON(&data); err != nil {
		context.JSON(200, []interface{}{nil, err.Error()})
		return
	}

	uuid, ok := data["uuid"].(float64)
	if !ok {
		context.JSON(200, []interface{}{nil, "invalid uuid type"})
		return
	}

	state, err := communication.ProcessIpc.GetProcessStatus(int(uuid))

	if err != nil {
		context.JSON(200, []interface{}{nil, *err})
		return
	}

	context.JSON(200, []interface{}{*state, nil})
}

func (p *Process) GetProcessOutput(context *gin.Context) {
	upgrader := websocket.Upgrader{
		ReadBufferSize:  1024,
		WriteBufferSize: 1024,
		CheckOrigin: func(r *http.Request) bool {
			return true
		},
	}

	var err error
	p.ws, err = upgrader.Upgrade(context.Writer, context.Request, nil)
	if err != nil {
		vlogger.AppLogger.Error(err.Error())
		return
	}

	go func() {
		defer p.ws.Close()
		for {
			if _, _, err := p.ws.ReadMessage(); err != nil {
				p.ws = nil
				return
			}
		}
	}()
}

func (p *Process) WriteProcessOutput(uuid int, output string) {
	if p.ws == nil {
		return
	}

	message := map[string]interface{}{
		"uuid": uuid,
		"data": output,
	}

	if err := p.ws.WriteJSON(message); err != nil {
		p.ws = nil
	}
}
